sympy/matrices/common.py

```diff
@@ -2989,6 +2991,46 @@ def __iter__(self): # supports numpy.matrix and numpy.array
         return iter(sympify(mat[r, c]) for r in range(self.rows) for c in range(cols))
 
 
+class MatrixKind(Kind):
+    """
+    Kind for all matrices in SymPy.
+
+    Basic class for this kind is ``MatrixBase`` and ``MatrixExpr``,
+    but any expression representing the matrix can have this.
+
+    Parameters
+    ==========
+
+    element_kind : Kind
+        Kind of the element. Default is ``NumberKind``, which means that
+        the matrix contains only numbers.
+
+    Examples
+    ========
+
+    >>> from sympy import MatrixSymbol, Integral
+    >>> from sympy.abc import x
+    >>> A = MatrixSymbol('A', 2,2)
+    >>> A.kind
+    MatrixKind(NumberKind)
+    >>> Integral(A,x).kind
+    MatrixKind(NumberKind)
```

Comment:
I think this should be expanded to something that makes it clear what the difference is between checking `isinstance(obj, Matrix)` and checking `obj.kind == MatrixKind(NumberKind)`. There should be text explaining when and why you should check the kind. Also how would you check for `MatrixKind` without specifying the element kind?